ui/components/src/TimeChart/TimeChart.tsx

```diff
@@ -270,61 +269,99 @@ export const TimeChart = forwardRef<ChartInstance, TimeChartProps>(function Time
+        const isControlKeyPressed = e.ctrlKey || e.metaKey;
+        if (isControlKeyPressed) {
+          e.preventDefault();
+        }
+
         // Determine where on chart canvas to plot pinned crosshair as markLine.
         const pointInGrid = getPointInGrid(e.nativeEvent.offsetX, e.nativeEvent.offsetY, chartRef.current);
         if (pointInGrid === null) {
           return;
         }
 
-        // Clear previously set pinned crosshair
-        const isCrosshairPinned = seriesMapping[seriesMapping.length - 1]?.name === PINNED_CROSSHAIR_SERIES_NAME;
-        if (tooltipPinnedCoords !== null && isCrosshairPinned) {
-          seriesMapping.pop();
-        } else if (seriesMapping.length !== data.length + 1) {
-          // Only add pinned crosshair line series when there is not one already in seriesMapping.
-          const pinnedCrosshair = merge(DEFAULT_PINNED_CROSSHAIR, {
-            markLine: {
-              data: [
-                {
-                  xAxis: pointInGrid[0],
-                },
-              ],
+        // Pin and unpin when clicking on chart canvas but not tooltip text.
+        if (isPinningEnabled && e.target instanceof HTMLCanvasElement) {
+          // Pin tooltip and update shared charts context to remember these coordinates.
+          const pinnedPos: CursorCoordinates = {
+            page: {
+              x: e.pageX,
+              y: e.pageY,
             },
-          });
-          seriesMapping.push(pinnedCrosshair);
-        }
+            client: {
+              x: e.clientX,
+              y: e.clientY,
+            },
+            plotCanvas: {
+              x: e.nativeEvent.offsetX,
+              y: e.nativeEvent.offsetY,
+            },
+            target: e.target,
+          };
 
-        // Pin and unpin when clicking on chart canvas but not tooltip text.
-        if (tooltipConfig.enablePinning && e.target instanceof HTMLCanvasElement) {
           setTooltipPinnedCoords((current) => {
             if (current === null) {
-              return {
-                page: {
-                  x: e.pageX,
-                  y: e.pageY,
-                },
-                client: {
-                  x: e.clientX,
-                  y: e.clientY,
-                },
-                plotCanvas: {
-                  x: e.nativeEvent.offsetX,
-                  y: e.nativeEvent.offsetY,
+              return pinnedPos;
+            } else {
+              setPinnedCrosshair(null);
+              return null;
+            }
+          });
+
+          setPinnedCrosshair((current) => {
+            // Only add pinned crosshair line series when there is not one already in seriesMapping.
+            if (current === null) {
+              const cursorX = pointInGrid[0];
+              const closestTimestamp = getClosestTimestampInFullDataset(data, cursorX);
+              const pinnedCrosshair = merge(DEFAULT_PINNED_CROSSHAIR, {
```

Comment:
Are you intending to mutate `DEFAULT_PINNED_CROSSHAIR` here (guessing not, but wanted to check since [merge](https://lodash.com/docs/4.17.15#merge) will do so)? If not, setting the first arg to an empty object would likely be safer.

Reply:
Nah it wasn't intentional, thanks for catching this, updated now!